tests/test_shardformer/test_layer/test_ring_attn.py

```diff
@@ -161,12 +164,12 @@ def check_packed_seq(seqlen, bs, nheads, d, dtype):
 def launch_single_ring(rank, world_size, port):
     colossalai.launch(rank, world_size, "localhost", port)
     check_packed_seq()
-    check_ring_attn()
+    check_ring_attn(sp_size=world_size)
 
 
 def launch_double_ring(rank, world_size, port):
     colossalai.launch(rank, world_size, "localhost", port)
-    check_ring_attn()
```

Comment:
Perhaps add a single test for the complext rank generation logic(https://github.com/hpcaitech/ColossalAI/blob/703bb5c18dd3a701209017de828e9cdc8ca58356/colossalai/shardformer/layer/attn.py#L483) for `world_size = 16`?
e.g. by putting rank generation into a single function `def get_double_ring_ranks(world_size, tp_size, sp_size, inner_ring_size, dp_pp_size)` and group creation after it.
Up to you🙂

Reply:
As the ci machine only have 8 gpus so I set world size = 8 to test the tp+sp config for the 2d-ring. Multiple nodes have already been tested locally and showed in the description of this PR below.

Reply:
Sure. It's possible to test the rank creation logic alone without using any GPU in a `world_size=16` example--in case someone messes it up later--but not adding a test case is also fine.

Reply:
What do you mean 'only rank creation logic'?

Reply:
put the for loops in a function 
```
def get_double_ring_ranks(world_size, tp_size, sp_size, inner_ring_size, dp_pp_size) -> Tuple[List[int]]:
        if tp_size > 1:
            # Create inner ring groups
            ranks = []
            for i in range(num_rings):
                start = i * num_ring_size
                end = (i + 1) * num_ring_size
                for idx in range(start, end):
                    inner_rank = []
                    for k in range(inner_ring_size):
                        current_num = idx + k * tp_size
                        if current_num >= end:
                            break
                        inner_rank.append(current_num)
                    if len(inner_rank) == inner_ring_size and inner_rank not in ranks:
                        ranks.append(inner_rank)
....
```

Reply:
CI is for not only rank creation but also accuracy test, and there are many rank test cases that I'm afraid one or two rank creation test is not enough. There are four variables impact the rank creation. So I test locally and put in the description.